lib/Sema/SemaDecl.cpp

```diff
@@ -10876,15 +10876,121 @@ void Sema::ActOnFinishKNRParamDeclarations(Scope *S, Declarator &D,
   }
 }
 
-/// ActOnBoundsExpr: attach a bounds expression to a parameter declaration.
-void Sema::ActOnBoundsExpr(DeclaratorDecl *D, BoundsExpr *Expr) {
+// ActOnBoundsDecl: handle a Checked C bounds declaration for a declarator.
+// Check that the declarator has the right kind of type for the bounds
+// expression and attach the bounds expression to the declarator.
+//
+// If checkReturnBounds is true, check the return bounds declaration
+// for a function declarator.  The typechecking logic is the same, but
+// the error messages need are slightly different.
+void Sema::ActOnBoundsDecl(DeclaratorDecl *D, BoundsExpr *Expr,
+                           bool checkReturnBounds) {
   if (!D || !Expr)
     return;
 
-  D->setBoundsExpr(Expr);
+  // If the bounds expression is invalid, skip type checking the declaration.
+  // TODO: add simple predicate to BoundsExpr that checks whether
+  // this is a valid bounds expression.
+  if (NullaryBoundsExpr *NB = dyn_cast<NullaryBoundsExpr>(Expr))
+    if (NB->getKind() == NullaryBoundsExpr::Invalid)
+      return;
+
+  unsigned DiagId = 0;
+  QualType Ty = D->getType();
+
+  // If we are checking the return bounds, get the function return type.
+  // If the type for the declaration wasn't a function type, bail out.
+  if (checkReturnBounds) {
+    assert(Ty->isFunctionType());
+    const FunctionType *FuncTy = Ty->getAs<FunctionType>();
+    if (FuncTy)
+      Ty = FuncTy->getReturnType();
+    else
+      return;
+  }
```

Comment:
I think a more LLVM-idiomatic way to do this is by putting the `getAs` call as a predicate on an `if` statement. Although I don't really know what the code around this does, because the style guide says one thing and much of the existing code does something else.